### PR TITLE
[ENT-468] Add support for Django 1.11.

### PR DESCRIPTION
**Pull request by UmanShahzad**
## TODO/NOTES

- Tests are passing.
- There were several warnings meant for Django 2.0 deprecation -- they were simple (but very tedious) changes so I got them through in this one, so we don't have to do them later.

## INFO

**Description:** This PR adds support for Django 1.11.

**JIRA:** [ENT-468](https://openedx.atlassian.net/browse/ENT-468)

**Dependencies:** None.

**Merge deadline:** Before the weekend since this is posted.

**Testing instructions:**

1. Run tox tests. (or see Travis below :) )

**Merge checklist:**

- [x] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @UmanShahzad! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-enterprise&number=143"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

**Comment by UmanShahzad** [comment]
Sorry for the current test failures -- it's due to a changed error message from Django's side, which is hard-coded into our tests...

File: enterprise/decorators.py [hunk 1]
```diff
@@ -139,7 +139,7 @@ def get(self, request, enterprise_uuid):
     @wraps(view)
     def wrapper(request, *args, **kwargs):
         """
-        Function wrapper.
+        Wrap the function.
```
Comment (UmanShahzad) [inline enterprise/decorators.py]: Yes, this was actually necessary. I know right.

File: README.rst [hunk 2]
```diff
@@ -13,8 +13,8 @@ enterprise
     :target: http://codecov.io/github/edx/edx-enterprise?branch=master
     :alt: Codecov
 
-.. image:: http://edx-enterprise.readthedocs.io/en/latest/?badge=latest
-    :target: http://edx-enterprise.readthedocs.io/en/latest/
+.. image:: http://open-edx-enterprise-service-documentation.readthedocs.io/en/latest/?badge=latest
+    :target: http://open-edx-enterprise-service-documentation.readthedocs.io/en/latest/
```
Comment (UmanShahzad) [inline README.rst]: Previously broken link. Since I'm updating a bunch of stuff, thought I'd touch this.

File: docs/testing.rst [hunk 3]
```diff
@@ -47,7 +47,8 @@ To run all tests under certain python and django versions:
 
 .. code-block:: bash
 
-    $ tox -e py27-django10          # run all tests under python 2.7 and django 1.10
+    $ tox -e py27-django110         # run all tests under python 2.7 and django 1.10
+    $ tox -e py35-django111         # run all tests under python 3.5 and django 1.11
```
Comment (UmanShahzad) [inline docs/testing.rst]: It was previously incorrect, and thought an extra example with py35 would be nice.

File: enterprise/__init__.py [hunk 4]
```diff
@@ -4,6 +4,6 @@
 
 from __future__ import absolute_import, unicode_literals
 
-__version__ = "0.36.11"
+__version__ = "0.37.0"
```
Comment (UmanShahzad) [inline enterprise/__init__.py]: When will this launch 1.0?

File: tests/test_views.py [hunk 5]
```diff
@@ -333,17 +332,12 @@ def test_post_patch_real_social_auth_enforced(
...
             client = Client()
             session = client.session
             session['partial_pipeline_token'] = True
             session.save()
             client.post(self.url)
-        expected = (
-            'Reverse for \'dashboard\' with arguments \'()\' and keyword '
-            'arguments \'{}\' not found. 0 pattern(s) tried: []'
-        )
-        assert str(excinfo.value) == expected
```
Comment (UmanShahzad) [inline tests/test_views.py]: This error message is no longer guaranteed -- different types of messages may be output. In the case where arguments and keywords are empty (as this was expecting), this is no longer the expected error message. Decided to just scratch hardcoding them throughout the codebase and look more importantly for the assertion.

File: tox.ini [hunk 6]
```diff
@@ -26,9 +31,10 @@ deps =
     django18: django>=1.8,<1.9
     django19: django>=1.9,<1.10
     django110: django>=1.10,<1.11
+    django111: django>=1.11,<2.0
     -r{toxinidir}/requirements/test.txt
 commands =
-    py.test {posargs}
+    py.test -Wd {posargs}
```
Comment (UmanShahzad) [inline tox.ini]: Default warning output is deprecated so we are recommended to be explicit in order to help us prepare for Django 2.0.

File: enterprise/migrations/0002_enterprisecustomerbrandingconfiguration.py [hunk 7]
```diff
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
```
Comment (UmanShahzad) [inline enterprise/migrations/0002_enterprisecustomerbrandingconfiguration.py]: All the changes that add this `on_delete` are harmless -- it's already the default. But a Django 2.0 deprecation notice said that we should add this because it will become a required argument. This was extremely tedious but I got it over with.

File: enterprise/models.py [hunk 8]
```diff
@@ -96,7 +96,11 @@ class Meta:
     )
```
Comment (UmanShahzad) [inline enterprise/models.py]: No migrations were made because, as noted above, it's already the default.

**Comment by UmanShahzad** [comment]
@haikuginger This is ready for a review.

File: test_settings.py [hunk 9]
```diff
@@ -115,6 +115,7 @@ def root(*args):
 
 # URL for the server that django client listens to by default.
 TEST_SERVER = "http://testserver"
+ALLOWED_HOSTS = ["testserver.enterprise"]
```
Comment (UmanShahzad) [inline test_settings.py]: > Changed in Django 1.11:
> In older versions, ALLOWED_HOSTS wasn’t checked when running tests.
> In older versions, ALLOWED_HOSTS wasn’t checked if DEBUG=True. This was also changed in Django 1.10.3, 1.9.11, and 1.8.16 to prevent a DNS rebinding attack.

File: tox.ini [hunk 10]
```diff
@@ -13,7 +13,12 @@ max-line-length = 120
 ; D200 = One-line docstring should fit on one line with quotes
 ; D203 = 1 blank line required before class docstring
 ; D212 = Multi-line docstring summary should start at the first line
-ignore = D101,D200,D203,D212
+; D406 = Section name should end with a newline
+; D407 = Missing dashed underline after section
+; D411 = Missing blank line before section
+; D412 = No blank lines allowed between a section header and its content
+; D413 = Missing blank line after last section
+ignore = D101,D200,D203,D212,D406,D407,D411,D412,D413
```
Comment (UmanShahzad) [inline tox.ini]: Decided to _not_ go and change a million comments to abide by these.

File: requirements/test.in [hunk 11]
```diff
@@ -8,4 +8,4 @@ mock                      # mocking library
 pytest-catchlog           # Show log output for test failures
 pytest-cov                # pytest extension for code coverage statistics
 pytest-django             # pytest extension for better Django support
-responses                 # utility library for mocking out the requests library
+responses                 # utility library for mocking out the requests library
```
Comment (haikuginger) [inline requirements/test.in]: Newline.

File: setup.py [hunk 12]
```diff
@@ -49,11 +49,11 @@ def get_version(*file_paths):
     ],
     include_package_data=True,
     install_requires=[
-        "Django>=1.8.15,<1.11",
+        "Django>=1.8.15,<2.0",
```
Comment (haikuginger) [inline setup.py]: As above with `<1.12`.

File: tests/test_tpa_pipeline.py [hunk 13]
```diff
@@ -112,9 +112,8 @@ def test_get_ec_for_request(self, fake_pipeline):
         with mock.patch('enterprise.tpa_pipeline.Registry') as fake_registry:
             fake_registry.get_from_pipeline.return_value = fake_provider
             assert get_enterprise_customer_for_request(request) == self.customer
-        with raises(NotConnectedToOpenEdX) as excinfo:
+        with self.assertRaises(NotConnectedToOpenEdX):
```
Comment (haikuginger) [inline tests/test_tpa_pipeline.py]: What's the rationale for switching away from the pytest style?

File: tox.ini [hunk 14]
```diff
@@ -26,9 +31,10 @@ deps =
     django18: django>=1.8,<1.9
     django19: django>=1.9,<1.10
     django110: django>=1.10,<1.11
+    django111: django>=1.11,<2.0
```
Comment (haikuginger) [inline tox.ini]: As above with the `<1.12` request.

File: enterprise/__init__.py (hunk 4 above)
Comment (UmanShahzad) [inline enterprise/__init__.py]: Makes sense.

File: enterprise/migrations/0002_enterprisecustomerbrandingconfiguration.py (hunk 7 above)
Comment (UmanShahzad) [inline enterprise/migrations/0002_enterprisecustomerbrandingconfiguration.py]: I didn't originally, no. I looked now and the only thing I could find about `on_delete` was https://docs.djangoproject.com/en/1.11/releases/1.9/#foreignkey-and-onetoonefield-on-delete-argument -- which is basically what I'm covering here.

File: requirements/base.in [hunk 15]
```diff
@@ -1,10 +1,10 @@
 # Core requirements for using this application
 
 celery                                  # Run task workers in other locations
-Django>=1.8,<1.11                       # Web application framework
+Django>=1.8,<2.0                        # Web application framework
```
Comment (UmanShahzad) [inline requirements/base.in]: Hmm, it may, but I'm worried it may have the negative effect of confusing someone who's well aware of the next planned version. If @brittneyexline agrees to make it `1.12`, I'll go ahead with that, not a big deal.

File: requirements/test.in (hunk 11 above)
Comment (UmanShahzad) [inline requirements/test.in]: Hehe, maybe the scripts in `make` should check for this (because they deleted them).

File: setup.py (hunk 12 above)
Comment (UmanShahzad) [inline setup.py]: If @brittneyexline thinks it well, I'll go for it. Not a problem.

File: tests/test_tpa_pipeline.py (hunk 13 above)
Comment (UmanShahzad) [inline tests/test_tpa_pipeline.py]: Leads us to have to (often) maintain explicitly hardcoded error messages in the tests. In fact, I am the one who added the `self.assertIsNotNone(excinfo.value)`, since it was previously a straight up string that was causing issues when error message results changed. That's a waste of time and I don't think there's any difference in detecting the exception vs. seeing if the error message is empty -- I've never run into empty error messages.

**Comment by UmanShahzad** [comment]
@haikuginger Thanks for the plus! I've replied to your comments -- as per the `<1.12` limiter, I personally don't have any major issue with that, but I want to get @brittneyexline to review it and let me know if that is indeed something we should change (just in case she feels otherwise and I have to go back and change it again).

File: requirements/base.in (hunk 15 above)
Comment (brittneyexline) [inline requirements/base.in]: I actually prefer to have it be 2.0 instead of 1.12, since 1.12 doesn't actually exist, and 2.0 is the next release version.

Comment (haikuginger) [inline requirements/base.in]: 2.0 works for me, then!

**Comment by UmanShahzad** [comment]
@haikuginger @brittneyexline Thanks, guys. Let me just add that newline and we should be good to go.

**Comment by haikuginger** [comment]
@UmanShahzad, I think I might not actually have tested the migration state properly - please stand by a bit.

**Comment by haikuginger** [comment]
Yeah, we've got some dependency hell going on. I may need to take this over to handle it.

**Comment by douglashall** [comment]
@haikuginger What issues are you running into with this one?

**Comment by haikuginger** [comment]
@douglashall, basically, we've got severe feature drift between what we've got in edx-platform and what we're using here. In one case, we've got a migration that adds a field that isn't present when we're using the version of a package that's installed in edx-platform. In another, we've got a dependency that, if installed, breaks edx-platform altogether. What's more, our setup.py file isn't using our actual requirements files during installation, so despite our fancy pip-compile scheme, we're not actually taking advantage of any of it. It's a mess, and while it could probably be worked around, it seems like the right thing to do to fix everything up before upgrading the Django version.

**Comment by UmanShahzad** [comment]
@haikuginger That... sounds pretty worrisome :)

**Comment by brittneyexline** [comment]
@haikuginger it seems like most of the problems with this is potential bad interactions when this is installed in the context of edx-enterprise - we've ran into issues before where tests pass in our repo's test suite and in edx-platform tests but then break in a live environment... I'm curious to hear about what approach you want to take to address this testing gap?

**Comment by douglashall** [comment]
@haikuginger It is unclear to me how this mess is manifesting itself. Could you elaborate?

**Comment by haikuginger** [comment]
@douglashall, for example, I have a devstack with an up-to-date master branch, and with all requirements installed. I then do `pip install -e /edx/src/edx-enterprise`, which overrides a bunch of the edx-platform requirements. But not all of them, you understand - only the ones explicitly specified in `setup.py`.

So, when I'm working in my development environment, after manually installing my local copy of enterprise, that's a different dependency environment than I'd get on a production system that prioritizes the dependencies in the edx-platform requirements files. And that's yet another dependency environment than is present while CI tests are running, because the tests explicitly install everything in the requirements.txt files.

One example is the current state of `django-simple-history`. [In setup.py](https://github.com/edx/edx-enterprise/blob/master/setup.py#L59), we just specify that it's present, so my development environment depends entirely on what's in my devstack - which, based on the edx-platform requirement, is 1.6.3. However, if someone were to manually upgrade that package in their devstack to the one specified by our own requirements files (1.9.0), [then when they ran `makemigrations`](https://github.com/edx/edx-enterprise/pull/137), they might have a migration created that would be nonsensical in a production environment running 1.6.3. And, our own tests, because they install the requirements from the requirements.txt files, wouldn't pick up anything wrong - in fact, if we were to make them check for missing migrations, they'd yell at us if we _didn't_ put that migration in place! But if I, with my 1.6.3 version, were to add a new database field somewhere, and run the `makemigrations` call from the LMS, a brand-new migration would be created to _revert_ the migration needed by 1.9.0.

Another example is `django-extensions`; edx-platform depends on 1.5.9, our requirements file calls for 1.7.9 (which causes Studio to mildly break when installed), and installing via setup.py grabs 1.8.0 , since we only specify `>=1.7.4`, and newer versions are preferred - that one causes the entire stack to go down, since 1.8.0 removes the UUIDField extension that `edx-submissions` relies on to work at all.

Basically, what we need to do is blow this whole pip-compile thing up, figure out what our immediate dependencies are, what _our_ useful ranges for those dependencies are, pin against the versions used in edx-platform wherever possible, and then direct `setup.py` to use that new requirements file directly, so that the dependencies in in development, testing, and deployment are, if not the same, at least as close as possible.

Another option is to take the whole thing and drop it straight into the edx-platform repo and never have to worry about dependency resolution ever again.

**Comment by UmanShahzad** [comment]
@haikuginger I understand that this is dependency hell, but mostly so when looking at it through the perspective of the edx-platform. Why should a satellite repository care about changing its own requirements (even when it's already trying to target a large range of them), to satisfy some explicit external entity? (I know the platform isn't _really_ external, but when you consider the goal of this repository to be a Django app not necessarily tied to any specific bigger application, then for that intent and purpose, we can consider it external).

tl;dr I believe it should've been the job of whoever tries to bump the edx-platform's requirement of edx-enterprise to make up for these consistencies, not for edx-enterprise to adapt itself to the platform. That would

1. Allow this repository to evolve rapidly.
2. Force the push of incrementing edx-platform's requirements to match _it's_ dependencies.

Of course, when the repository explicitly tries to tie (and thus couple) itself so closely to the edx-platform, then, as you said, it would be better off just being a part of its code.

**Comment by haikuginger** [comment]
Aha. I think I've figured out the right solution for the dependency problem. I'll be pushing a commit here tonight; it'll need reviews, obviously.

**Comment by haikuginger** [comment]
Short explanation of what I've done here (will update more later):

We want our testing environment (and our development environment) to reflect what's present in edx-platform; at the moment, Enterprise is embedded deeply, without an easy way out.

Therefore, I've created edx release-focused test branches; one each for Ficus, Ginkgo, the current state of the master branch, as well as a speculative Hawthorn test suite, which includes Django 1.11.

The different dependency environments should give us a good range of testing for different edx-platform use cases. As for installation, `setup.py` will now only attempt to install packages that don't come for free from `edx-platform`, preserving the dependency state of a devstack similar to what would be available on a full instance deployed using that version of master.

When we go to add this version of edx-enterprise to edx-platform, we'll also need to update a few dependencies there. I've looked through, and don't anticipate any issues:

- `unicodecsv` from 0.9.3 to 0.14.1
- `django-config-models` from 0.1.3 to 0.1.8
- `django-simple-history` from 1.6.3 to 1.9.0

The current version of `unicodecsv` used on master is syntactically invalid in python3, making it impossible to test in python3 with it.

The current version of `django-config-models` has a version conflict with the current version of `djangorestframework`.

An upgrade of `django-simple-history` for Django 1.11 will be necessary anyway, and we've already (if mistakenly) generated our own migrations for it. Rather than revert those migrations, only to generate them again later, we should bring edx-platform up-to-date.

Pinging @itsjeyd for review on this change.

File: requirements/base.in [hunk 16]
```diff
@@ -1,19 +1,4 @@
...
-Pillow                                  # Image manipulation module, required to use ImageField
-django_extensions                       # Required to use TimeStampedModel
-django-simple-history                   # History for Django models
-django-object-actions                   # Object actions in Django admin
-edx-rest-api-client                     # For accessing the Enrollment API (and possibly other edX APIs)
-django-config-models==0.1.6
-requests                                # Required for SAPSuccessFactorsAPIClient
-django-waffle                           # Allows ability to add and control flags and switches for features
+django-object-actions==0.10.0           # Object actions in Django admin
+django-waffle==0.12.0                   # Allows ability to add and control flags and switches for features
```
Comment (brittneyexline) [inline requirements/base.in]: i'm pretty sure django-waffle is in [edx-platform](https://github.com/edx/edx-platform/blob/master/requirements/edx/base.txt#L43), so should this be here?

Comment (haikuginger) [inline requirements/base.in]: It isn't in edx-platform for Ficus.

Comment (brittneyexline) [inline requirements/base.in]: got it, thanks

**Comment by haikuginger** [comment]
@brittneyexline, I'll be doing the upgrades for the three dependencies I mentioned as part of the version bump for this PR, but for the others, I advocate a "wait and see where Hawthorn lands" approach, along with updating the requirements for the `test-master.in` file as we see changes. I've updated the PR template as well to reflect the additional step of checking requirements against edx-platform to make sure our tests are valid.

File: requirements/test-ficus.in [hunk 17]
```diff
@@ -0,0 +1,22 @@
+# Deviations:
+# In edx-platform, DRF is a custom version; we do not use the extra features.
+# In edx-platform, django-config-models is 0.1.3; 0.1.5 has no substantive
+# behavioral changes, but explicitly supports DRF 3.2.3.
+# In edx-platform, unicodecsv is at 0.9.4, which is syntactically
+# invalid for testing in Python 3
```
Comment (itsjeyd) [inline requirements/test-ficus.in]: @haikuginger For readability, can we do:

```
# Deviations:
#
# - In edx-platform, DRF is a custom version; we do not use the extra features.
# - In edx-platform, django-config-models is 0.1.3; 0.1.5 has no substantive behavioral changes,
#   but explicitly supports DRF 3.2.3.
# - In edx-platform, unicodecsv is at 0.9.4, which is syntactically invalid for testing in Python 3.
```

File: requirements/test-ginkgo.in [hunk 18]
```diff
@@ -0,0 +1,23 @@
+# Deviations:
+# In edx-platform, DRF is a custom version; we do not use the extra features.
+# In edx-platform, django-config-models is 0.1.3; 0.1.5 has no substantive
+# behavioral changes, but explicitly supports DRF 3.2.3.
+# In edx-platform, unicodecsv is at 0.9.4, which is syntactically
+# invalid for testing in Python 3
```
Comment (itsjeyd) [inline requirements/test-ginkgo.in]: @haikuginger Similar comment here; let's do:

```
# Deviations:
#
# - In edx-platform, DRF is a custom version; we do not use the extra features.
# - In edx-platform, django-config-models is 0.1.3; 0.1.5 has no substantive behavioral changes,
#   but explicitly supports DRF 3.2.3.
# - In edx-platform, unicodecsv is at 0.9.4, which is syntactically invalid for testing in Python 3.
```

File: requirements/test-master.in [hunk 19]
```diff
@@ -0,0 +1,19 @@
+# Deviations:
+# In edx-platform, DRF is a custom version; we do not use the extra features.
```
Comment (itsjeyd) [inline requirements/test-master.in]: @haikuginger To prepare for additional items, let's do:

```
# Deviations:
#
# - In edx-platform, DRF is a custom version; we do not use the extra features.
```
